### ISSUE TITLE
Is makeFailsafe redundant in ScopeGuard?

### ISSUE THREAD

**Issue by ruiyuanlu** [facebook/folly#1996]
Hi, I'm new to C++, I wonder if makeFailsafe [here](https://github.com/facebook/folly/blob/v2023.05.08.00/folly/ScopeGuard.h#L160-L175) is redundant in ScopeGuard. I think it actually does nothing. 

Is it OK to remove these when used in my project? What does failsafe actually do?

**Comment by r00tk1ts** [comment]
If the functor passed in has a nothrow(false) copy/move constructor, the ScopedGuard constructor would throw exception itself.  The overloaded `makeFailsafe` functions use the tag-dispatch tech to deal with these cases. When you pass a nothrow(false) version, it will internally make a ScopeGuard object by deducing Fn to std::ref(*fn). As the std::refference_type has both nothrow(true) copy/move constructor, so the statement `function_(std::forward<Fn>(fn))` will not throw any exceptions.

For example:
```cpp
class functor{
public:
    functor() = default; 

    functor(const functor &other) noexcept(false)  {
        std::cout<<"copy construct"<<std::endl;
        throw std::runtime_error("oops! ...");
    }

    void operator() () {
        std::cout<<"exit called"<<std::endl;
    }
};

int main() {
    auto guard = folly::makeGuard(functor());

    return 0;
}
```

While our functor has a noexcept(false) copy ctor, and it will surely throw a runtime_error exception in our code.
If we remove makeFailsafe in ScopeGuard impl, the program will be terminated due to the runtime_error exception. 
But now, the program will surely get "exit called" printing msg.

In short, makeFailsafe is used to dispatch different nothrow(true or false) ctor version for passing functor. If the functor's ctor is nothrow(true), we stored a copied functor in function_ class member, if not we stored a copied std::ref(&functor) in function_.

**Comment by ruiyuanlu** [comment]
I see.

One more question about the failsafe execution timing. 

The failsafe is r-value right?

 I mean in this case, failsafe is invoked when ScopeGuardImpl throw error and exit right？which differs from the usual timing right ?

**Comment by r00tk1ts** [comment]
> I see.
> 
> One more question about the failsafe execution timing.
> 
> The failsafe is r-value right?
> 
> I mean in this case, failsafe is invoked when ScopeGuardImpl throw error and exit right？which differs from the usual timing right ?

You are correct.

**Comment by ruiyuanlu** [comment]
Thx for your explain~
